packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -627,16 +644,57 @@ export function allocateChildren(vnode: VCustomElement, vm: VM) {
             );
         }
     }
+
+    // If any of the children being allocated are VFragments, we remove the text delimiters and flatten all immediate
+    // children VFragments to avoid them interfering with default slot behavior.
+    const allocatedChildren = flattenFragmentsInChildren(children);
+    vnode.children = allocatedChildren;
+    vm.aChildren = allocatedChildren;
+
     if (shadowMode === ShadowMode.Synthetic || renderMode === RenderMode.Light) {
         // slow path
-        allocateInSlot(vm, children, vnode.owner);
+        allocateInSlot(vm, allocatedChildren, vnode.owner);
         // save the allocated children in case this vnode is reused.
-        vnode.aChildren = children;
+        vnode.aChildren = allocatedChildren;
         // every child vnode is now allocated, and the host should receive none directly, it receives them via the shadow!
         vnode.children = EmptyArray;
     }
 }
 
+/**
+ * Flattens the contents of all VFragments in an array of VNodes, removes the text delimiters on those VFragments, and
+ * marks the resulting children array as dynamic.
+ *
+ * With the delimiters removed, the contents must be treated as dynamic for them to rerender correctly.
+ *
+ * This function is used for slotted VFragments to avoid the text delimiters interfering with slotting functionality.
+ */
+function flattenFragmentsInChildren(children: readonly (VNode | null)[]): (VNode | null)[] {
+    const flattenedChildren: (VNode | null)[] = [];
+    const nodeStack: (VNode | null)[] = [];
+
+    // First iteration is slightly different than the rest of the traversal
+    for (let i = children.length - 1; i > -1; i -= 1) {
+        nodeStack.push(children[i]);
+    }
+
+    while (nodeStack.length > 0) {
+        const currentNode = nodeStack.pop();
+        if (currentNode && isVFragment(currentNode)) {
+            const fChildren = currentNode.children;
+            // Ignore the start and end text node delimiters
+            for (let i = fChildren.length - 2; i > 0; i -= 1) {
+                nodeStack.push(fChildren[i]);
+            }
+        } else {
+            flattenedChildren.push(currentNode!);
```

Comment:
To avoid this `!` (and the `isUndefined()` check I suggest above), we could change the for-loop to:

```js
    let currentNode
    while (!isUndefined(currentNode = nodeStack.pop())) {
```

I know some folks don't like this pattern, but it does make TypeScript happy. And it avoids the `.length` check. 🤷